### ISSUE TITLE
Load order conflict with Angular's NgZone

### ISSUE THREAD

**Issue by bkbooth** [TrackJs/trackjs-package#15]
We recently added TrackJS to an Angular 2 project and found that in some cases the view wasn't updating after HTTP requests. The HTTP success/fail handlers didn't appear to be running inside Angular's `NgZone`. We were able to confirm this by either injecting the `NgZone` service and wrapping the handlers' logic in `ngZone.run(...)`, or injecting the `ChangeDetectorRef` and running `changeDetector.detectChanges()`, both of which fixed the issue. We're using Angular's `Http` service, so it definitely should be running the handlers inside `NgZone`, we shouldn't need to manually use `NgZone` or `ChangeDetectorRef` in this case.

Disabling the TrackJS tracker fixes the issue, but as far as we can tell this is a load order conflict with the TrackJS tracker because moving the tracker config and loader to the end of `<body>` instead of `<head>` also fixes the issue for us. We're using this as a temporary workaround but we'd ideally like to move the tracker back to `<head>` so that we can catch errors that might happen during Angular's initialisation.

**Comment by BrandesEric** [comment]
Thanks for letting us know!  I confess we're not Angular 2 experts so if you had any kind of simple reproduction case that exhibited the issue that would certainly help us get a fix out sooner :)

We do some wrapping of native code.  Based on your description I wonder if our wrapping of XmlHttpRequest is conflicting with NgZone?  There are two things that, barring a reproduction case, could help us narrow this down.

1) Would you be willing to try leaving the tracker in the `<head>` tag but adding the following config?

```
window._trackJs = {
  token: "YOUR_TOKEN",
  network: {
    enabled: false
  }
}
```

This will turn off our native wrapping of XmlHttpRequest, but leave the rest of the tracker functional.  I'd be curious to know if this fixes the issue?

2) If that doesn't do the trick, we do also instrument `addEventListener` which is potentially how success callbacks are subscribing?

If so, you can try a similar config:
 ```
window._trackJs = {
  token: "YOUR_TOKEN",
  callback: {
    enabled: false
  }
}
```
This will turn off our wrapping of the native addEventListener function.   

If you're able to try either of those, would you let me know if you see any improvement?

Thanks!
Eric

**Comment by bkbooth** [comment]
Hi Eric, thanks for the quick reply!

I'm still at work and I've had to move focus elsewhere but I'll try both of those and get back to you as soon as I can. Of the two suggestions, based on what you've said and what we're seeing on Chrome's dev tools my hunch is that the first option will probably fix the issue.

I'll also try and put together a minimal reproducible example, but that might take a little longer.

**Comment by bkbooth** [comment]
My hunch was wrong, with option 1 I still see the issue, but with option 2 the issue is gone.

**Comment by BrandesEric** [comment]
Very interesting!  And if I understood you correctly earlier - the issue is also not present if you let Angular 2 (and other dependencies) load first and load the tracker script second?

I will attempt to create a repro, but if you're able to come up with anything that will undoubtedly be more effective :)  If you're also just wanting to move on, you can leave that setting configured as-is, and my guess is you won't lose much in the way of context.  Usually you want to keep it on in order to get better stack traces, but in the case of Angular 2's unified error handling I suspect it won't make much difference.  

Did you get a chance to try our angular 2 integration snippet?   http://docs.trackjs.com/tracker/framework-integrations#angular-2

Anyways, I'll try and see if I can get a repro case, but if you're able to come up with one please let us know!

Thanks!
Eric

**Comment by bkbooth** [comment]
I managed to get a reproducible example setup for you: https://plnkr.co/edit/ESArNXsIOMSsfvLUIq8p?p=preview

If it's not obvious, the normal behaviour is for the "Loading people..." message to disappear and a list of names to appear when the HTTP request returns.

Let me know if you need any more help digging into this.

**Comment by BrandesEric** [comment]
Wow, this is perfect!  We will dig in on this and see what we can come up with :)

**Comment by bkbooth** [comment]
And in response to your other questions:

Yeah, if you load the tracker after Angular 2 and it's dependencies load then everything seems to work fine.

We haven't tried the integration snippet yet, but we definitely will.

Thanks again for your help and quick responses!

**Comment by BrandesEric** [comment]
@bkbooth I think I've gotten to the root of the issue but don't have a good solution for you yet.

Here is what I think is happening:

Both `NgZone` and our tracker script monkey-patch `addEventListener`.  The problem is that we patch at different levels.

[EventTarget](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget)  is the basis for all `addEventListener` functions.  That is, `XMLHttpRequest`, `Element`, and many other objects ultimately derive their `addEventListener` implementations from `EventTarget`
![image](https://cloud.githubusercontent.com/assets/640877/20986459/67b4b1fe-bc8e-11e6-85a0-3e41968965ba.png)

The problem is, if our tracker script loads first, we patch `XMLHttpRequest.prototype.addEventListener` with our custom wrapping function.  Our custom wrapping function ultimately delegates to the original, browser native `EventTarget.prototype.addEventListener` implementation.

You can think of it almost like this pseudoish code:

```
var originalNativeFunc = XMLHttpRequest.prototype.addEventListener // this is === EventTarget.prototype.addEventListener until NgZone loads
XMLHttpRequest.prototype.addEventListener = function (){
    // do our stuff
    originalNativeFunc.apply(this, args)
}
```

 Then, when `NgZone` loads, it patches `EventTarget.prototype.addEventListener`.  But the problem is, that change is not propagated to `XMLHttpRequest` since we already replaced that `addEventListener` function with our own custom one.

![image](https://cloud.githubusercontent.com/assets/640877/20986718/6a415426-bc8f-11e6-8729-e4f0f088fb01.png)


This is also why it works if you load our script second - `NgZone` has already done its work, so the `XMLHttpRequest.prototype.addEventListener` that we patch is ultimately the `NgZone` one.

Anyways, the two immediate solutions are to leave the callback wrapping off (what I would do) or load our script second.

The long term fix is maybe either for us to patch at `EventTarget.prototype.addEventListener`  or have our function always delegate to `EventTarget.prototype.addEventListener` rather than the copy of the `originalNativeFunc`.  Our resident JS expert is traveling at the moment though, so we'll need to have a discussion once he's back and then there will be much testing.  Just being realistic, I think we're looking at a couple of weeks before this goes live :(

**Comment by BrandesEric** [comment]
Oh, and for completeness, here is a fork of your excellent original Plnkr with the tracker script patching at the `EventTarget.prototype` level: https://plnkr.co/edit/XDVgheBzHsu98GcM1jES?p=preview

**Comment by bkbooth** [comment]
Thanks @BrandesEric! I really appreciate you taking the time to investigate and reply to this in detail. Take the time you need to discuss, resolve and test a long term fix, for now we'll work out which is the best of the two short-term solutions for us.

**Comment by toddhgardner** [comment]
patched with a workaround in 2.8.3
